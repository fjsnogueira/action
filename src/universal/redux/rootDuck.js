export const RESET = '@@root/RESET';

export default (appReducer) => (state, action) => {
  if (action && action.type === RESET) {
    const {whitelist} = action.payload;
    const newState = {};
    whitelist.forEach((keeper) => {
      if (keeper in state) {
        newState[keeper] = state[keeper];
      }
    });
    state = newState;
  }
  return appReducer(state, action);
};

export const reset = (whitelist = []) => ({
  type: RESET,
  payload: { whitelist }
});
