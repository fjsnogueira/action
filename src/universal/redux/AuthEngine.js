import {setAuthToken, removeAuthToken, DEFAULT_AUTH_REDUCER_NAME} from './authDuck';

export default class AuthEngine {
  constructor(store, reducerName = DEFAULT_AUTH_REDUCER_NAME) {
    this.store = store;
    this.reducerName = reducerName;
  }
  saveToken(name, token, options, callback) {
    this.store.dispatch(setAuthToken(token, this.reducerName));
    if (callback) callback(null, token);
  }

  removeToken(name, callback) {
    const token = this.store.getState()[this.reducerName].token;
    this.store.dispatch(removeAuthToken());
    if (callback) callback(null, token);
  }

  loadToken(name, callback) {
    const token = this.store.getState()[this.reducerName].token;
    if (callback) callback(null, token);
  }
}
