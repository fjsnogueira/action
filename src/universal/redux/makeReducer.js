import {combineReducers} from 'redux';
import {cashayReducer} from 'cashay';
import toasts from 'universal/modules/toast/ducks/toastDuck';
import {reducer as formReducer, actionTypes} from 'redux-form';
import {reducer as storageReducer} from 'redux-storage-whitelist-fn';
import storageMerger from 'universal/redux/storageMerger';
import makeRootReducer from 'universal/redux/rootDuck';
import menuReducer from 'universal/modules/menu/ducks/menuDuck';
import dashReducer from 'universal/modules/dashboard/ducks/dashDuck';
import auth, {DEFAULT_AUTH_REDUCER_NAME} from './authDuck';

const {SET_SUBMIT_SUCCEEDED} = actionTypes;

// wipe the value clean when submitted
const wipeAfterSuccess = (formToClear) => (state, action) => {
  if (action.type === SET_SUBMIT_SUCCEEDED) {
    if (action.meta.form === formToClear) {
      return undefined;
    }
  }
  return state;
};

const formPluginFactory = () => {
  // add new fields to this array if you want em cleared
  const clearMeAfterSubmit = ['agendaInput', 'inviteTeamMember'];
  return clearMeAfterSubmit.reduce((formPlugin, name) => {
    formPlugin[name] = wipeAfterSuccess(name);
    return formPlugin;
  }, {});
};

const formPlugin = formPluginFactory();

const appReducers = {
  [DEFAULT_AUTH_REDUCER_NAME]: auth,
  cashay: cashayReducer,
  form: formReducer.plugin(formPlugin),
  dash: dashReducer,
  menu: menuReducer,
  toasts,
};


export default (newReducers) => {
  Object.assign(appReducers, newReducers);
  const appReducer = combineReducers({...appReducers});
  const rootReducer = makeRootReducer(appReducer);
  return storageReducer(rootReducer, storageMerger);
};
