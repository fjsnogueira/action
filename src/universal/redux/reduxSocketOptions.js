import {cashay, Transport} from 'cashay';
import socketCluster from 'socketcluster-client';
import subscriber from 'universal/subscriptions/subscriber';
import AuthEngine from './AuthEngine';

const onConnect = (options, hocOptions, socket) => {
  if (!cashay.priorityTransport) {
    const sendToServer = (request) => {
      return new Promise((resolve) => {
        socket.emit('graphql', request, (error, response) => {
          resolve(response);
        });
      });
    };
    const priorityTransport = new Transport(sendToServer);
    cashay.create({priorityTransport, subscriber});
  }
};
const onDisconnect = () => {
  cashay.create({priorityTransport: null});
};
export default ({AuthEngine, socketCluster, onConnect, onDisconnect, keepAlive: 3000});
