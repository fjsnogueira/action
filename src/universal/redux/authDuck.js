import {cashay} from 'cashay';
import jwtDecode from 'jwt-decode';
import ActionHTTPTransport from '../utils/ActionHTTPTransport';
import {selectSegmentProfile} from 'universal/redux/segmentActions';
import raven from 'raven-js';

const SET_AUTH_TOKEN = '@@authToken/SET_AUTH_TOKEN';
const REMOVE_AUTH_TOKEN = '@@authToken/REMOVE_AUTH_TOKEN';
const SET_NEXT_URL = '@@authToken/SET_NEXT_URL';
const UNSET_NEXT_URL = '@@authToken/UNSET_NEXT_URL';

export const DEFAULT_AUTH_REDUCER_NAME = 'auth';

const initialState = {
  // next URL to navigate to after authenticating:
  nextUrl: null,
  // parsed jwt token:
  obj: {
    // nextUrl, rol, and tms are not guaranteed
    aud: null,
    exp: null,
    iat: null,
    iss: null,
    sub: null,
  },
  // Raw jwt token:
  token: null
};

export default function reducer(state = initialState, action = {type: ''}) {
  switch (action.type) {
    case SET_AUTH_TOKEN: {
      const {obj, token} = action.payload;
      return {
        ...state,
        obj,
        token
      };
    }
    case REMOVE_AUTH_TOKEN: {
      const {nextUrl} = state;
      return {
        ...initialState,
        nextUrl
      };
    }
    case SET_NEXT_URL: {
      const {nextUrl} = action.payload;
      return {
        ...state,
        nextUrl
      };
    }
    case UNSET_NEXT_URL: {
      const {obj, token} = state;
      return {
        ...initialState,
        obj,
        token
      };
    }
    default:
      return state;
  }
}

export function setAuthToken(authToken, reducerName = DEFAULT_AUTH_REDUCER_NAME) {
  return (dispatch, getState) => {
    if (!authToken) {
      throw new Error('setAuthToken action created with undefined authToken');
    }
    let obj = null;
    try {
      obj = jwtDecode(authToken);
    } catch (e) {
      throw new Error(`unable to decode jwt: ${e}`);
    }

    cashay.create({httpTransport: new ActionHTTPTransport(authToken)});
    if (typeof __PRODUCTION__ !== 'undefined' && __PRODUCTION__) {
      /*
       * Sentry error reporting meta-information. Raven object is set via SSR.
       * See server/Html.js for how this is initialized
       */
      const profile = selectSegmentProfile(getState(), reducerName);
      raven.setUserContext({
        id: obj.sub,
        email: profile.email
      });
    }
    dispatch({
      type: SET_AUTH_TOKEN,
      payload: {
        obj,
        token: authToken
      }
    });
  };
}

export function removeAuthToken() {
  return (dispatch) => {
    if (typeof __PRODUCTION__ !== 'undefined' && __PRODUCTION__) {
      /*
       * Sentry error reporting meta-information. Raven object is set via SSR.
       * See server/Html.js for how this is initialized
       */
      raven.setUserContext({});
    }
    dispatch({ type: REMOVE_AUTH_TOKEN });
  };
}

export function setNextUrl(nextUrl) {
  return (dispatch) => {
    dispatch({
      type: SET_NEXT_URL,
      payload: { nextUrl }
    });
  };
}

export function unsetNextUrl() {
  return (dispatch) => {
    dispatch({ type: UNSET_NEXT_URL });
  };
}
