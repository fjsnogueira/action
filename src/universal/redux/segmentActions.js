import {EventTypes} from 'redux-segment';
import {cashay} from 'cashay';
import {DEFAULT_AUTH_REDUCER_NAME} from './authDuck';
import {getAuthQueryString, getAuthedOptions} from './getAuthedUser';

/**
 * Sometimes, we just want to track events in segment.
 */

const SEGMENT_EVENT = '@@segment/EVENT';

const defaultProfile = {
  avatar: null,
  email: null,
  id: null,
  name: null
};

export function selectSegmentProfile(state, authReducer = DEFAULT_AUTH_REDUCER_NAME) {
  const userId = state[authReducer].obj.sub;
  if (!userId) {
    return defaultProfile;
  }
  const {user} = cashay.query(getAuthQueryString, getAuthedOptions(userId)).data;

  return ({
    avatar: user.picture,
    email: user.email,
    id: user.id,
    name: user.preferredName,
  });
}

export function segmentEventIdentify() {
  return (dispatch, getState) => {
    const profile = selectSegmentProfile(getState());
    dispatch({
      type: SEGMENT_EVENT,
      meta: {
        analytics: {
          eventType: EventTypes.identify,
          eventPayload: {
            userId: profile.id,
            traits: {
              avatar: profile.picture,
              email: profile.email,
              name: profile.name
            }
          }
        }
      }
    });
  };
}

export function segmentEventTrack(event, properties, options) {
  return (dispatch, getState) => {
    const profile = selectSegmentProfile(getState());
    const propertiesOut = Object.assign({}, profile, properties);
    const optionsOut = Object.assign({}, {context: {traits: profile}}, options);

    dispatch({
      type: SEGMENT_EVENT,
      meta: {
        analytics: {
          eventType: EventTypes.track,
          eventPayload: {
            event,
            properties: propertiesOut,
            options: optionsOut
          }
        }
      }
    });
  };
}

export function segmentEventPage(name, category, properties, options) {
  return (dispatch, getState) => {
    const profile = selectSegmentProfile(getState());
    const propertiesOut = Object.assign({}, profile, properties);
    const optionsOut = Object.assign({}, {context: {traits: profile}}, options);

    dispatch({
      type: SEGMENT_EVENT,
      meta: {
        analytics: {
          eventType: EventTypes.page,
          eventPayload: {
            name,
            category,
            properties: propertiesOut,
            options: optionsOut
          }
        }
      }
    });
  };
}
