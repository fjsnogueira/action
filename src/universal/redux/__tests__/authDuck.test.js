import raven from 'raven-js';
import {applyMiddleware, combineReducers, createStore} from 'redux';
import thunk from 'redux-thunk';
import auth, {setAuthToken, removeAuthToken, setNextUrl, unsetNextUrl} from '../authDuck';
import {testToken, testTokenData} from './testTokens';
import * as segmentActions from '../segmentActions';

raven.setUserContext = jest.fn();
segmentActions.selectSegmentTraits = jest.fn(() => ({email: 'a@a.co'}));
segmentActions.segmentEventIdentify = jest.fn(() => ({ type: '@@segment/EVENT' }));
segmentActions.segmentEventReset = jest.fn(() => ({ type: '@@segment/EVENT' }));
let appReducer;
let store;
let initialState;

beforeEach(() => {
  appReducer = combineReducers({auth});
  store = createStore(appReducer, {}, applyMiddleware(thunk));
  initialState = store.getState();
});


test('initial state', () => {
  expect(initialState).toMatchSnapshot();
});

test('can setAuthToken w/token decode', () => {
  global.__PRODUCTION__ = true;
  store.dispatch(setAuthToken(testToken));
  expect(raven.setUserContext).toBeCalledWith({id: testTokenData.sub, email: 'a@a.co'});
  expect(segmentActions.segmentEventIdentify).toBeCalled();
  expect(store.getState()).toMatchSnapshot();
});

test('throws when token undefined', () => {
  expect(() =>
    store.dispatch(setAuthToken(undefined))
  ).toThrow();
});

test('throws when token invalid', () => {
  expect(() =>
    store.dispatch(setAuthToken(42))
  ).toThrow();
});

test('can removeAuthToken', () => {
  store.dispatch(setAuthToken(testToken));
  store.dispatch(removeAuthToken());
  expect(segmentActions.segmentEventReset).toBeCalled();
  expect(store.getState()).toMatchSnapshot();
});

test('setNextUrl', () => {
  store.dispatch(setNextUrl('/foo'));
  expect(store.getState()).toMatchSnapshot();
});

test('unsetNextUrl', () => {
  store.dispatch(setNextUrl('/foo'));
  store.dispatch(unsetNextUrl());
  expect(store.getState()).toMatchSnapshot();
});
