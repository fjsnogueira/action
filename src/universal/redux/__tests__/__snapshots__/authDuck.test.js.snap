// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can removeAuthToken 1`] = `
Object {
  "auth": Object {
    "nextUrl": null,
    "obj": Object {
      "aud": null,
      "exp": null,
      "iat": null,
      "iss": null,
      "sub": null,
    },
    "token": null,
  },
}
`;

exports[`can setAuthToken w/token decode 1`] = `
Object {
  "auth": Object {
    "nextUrl": null,
    "obj": Object {
      "name": "Moe",
      "rol": "su",
      "sub": "1234567890",
      "tms": Array [
        "team123",
      ],
    },
    "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6Ik1vZSIsInJvbCI6InN1IiwidG1zIjpbInRlYW0xMjMiXX0.N0NvXB5JCkwwHUsir4hC5noyoWuAhrq6JRvGlvdww5Q",
  },
}
`;

exports[`initial state 1`] = `
Object {
  "auth": Object {
    "nextUrl": null,
    "obj": Object {
      "aud": null,
      "exp": null,
      "iat": null,
      "iss": null,
      "sub": null,
    },
    "token": null,
  },
}
`;

exports[`setNextUrl 1`] = `
Object {
  "auth": Object {
    "nextUrl": "/foo",
    "obj": Object {
      "aud": null,
      "exp": null,
      "iat": null,
      "iss": null,
      "sub": null,
    },
    "token": null,
  },
}
`;

exports[`unsetNextUrl 1`] = `
Object {
  "auth": Object {
    "nextUrl": null,
    "obj": Object {
      "aud": null,
      "exp": null,
      "iat": null,
      "iss": null,
      "sub": null,
    },
    "token": null,
  },
}
`;
